File: 2019_Haskell/Day1.hs
main :: IO ()
main = do
    numbers <- map read . lines <$> readFile "input/day_1.txt"
    print $ part1 numbers
    print $ part2 numbers

part1 :: [Int] -> Int
part1 = sum . map fuel

fuel :: Int -> Int
fuel 0 = 0
fuel n = n `div` 3 - 2

part2 :: [Int] -> Int
part2 = sum . map fuel'

fuel' :: Int -> Int
fuel' n | n > 0 = n' + fuel' n'
    where n' = fuel n
fuel' _ = 0

File: 2024_rust/src/bin/day_1.rs
use itertools::Itertools;

const INPUT: &str = include_str!("../../input/day_1.txt");

fn parse() -> (Vec<usize>, Vec<usize>) {
    INPUT
        .trim()
        .lines()
        .map(|line| {
            line.split_ascii_whitespace()
                .map(|n| n.parse::<usize>().unwrap())
                .collect_tuple()
                .unwrap()
        })
        .collect()
}

fn part1() -> usize {
    let (l, r) = parse();
    l.iter()
        .sorted()
        .zip(r.iter().sorted())
        .map(|(&a, &b)| a.abs_diff(b))
        .sum()
}

fn part2() -> usize {
    let (l, r) = parse();
    l.iter()
        .map(|a| a * r.iter().filter(|n| *n == a).count())
        .sum()
}

fn main() {
    println!("Part 1: {}", part1());
    println!("Part 2: {}", part2());
}
